### PR TITLE
Animation graph preview/variable manipulation

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: #

### Changelog

* Add some editor decicated APIs to support preview animation blend/animation transition in animation graphs, as well as more flexibly manipulate variables in animation graphs. 

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.

  > Manual trigger with `@cocos-robot run test cases` afterward.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: editor/src/marionette/preview.ts [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+
+    public destroy() {
+    }
+
+    public evaluate() {
+        this._blendBuffer.apply();
+    }
+
+    public addVariable(id: string, type: VariableType, value: Value) {
+        this._varInstances[id] = new VarInstance(type, value);
```
Comment (SantyWang) [inline editor/src/marionette/preview.ts]: What if add variable with same name twice?

File: editor/src/marionette/preview.ts [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    }
+
+    public removeVariable(id: string) {
+        delete this._varInstances[id];
+    }
+
+    public updateVariable(id: string, value: Value) {
+        this._varInstances[id].value = value;
```
Comment (SantyWang) [inline editor/src/marionette/preview.ts]: Need to check if variable exists?

File: editor/src/marionette/preview.ts [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+            this._timelineStatsDirty = false;
+        }
+        return this._timelineStats;
+    }
+
+    /**
+     * Gets an iterable to the weights of each motion(that has runtime ID).
+     */
+    public queryWeights(): Iterable<[RuntimeID, number]> {
+        const getWeightsRecursive = function* (motionEval: MotionEval, weight: number): Iterable<[RuntimeID, number]> {
```
Comment (SantyWang) [inline editor/src/marionette/preview.ts:94]: What does function* mean?

File: editor/src/marionette/preview.ts [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+        return [];
+    }
+
+    public setMotion(motion: Motion) {
+        this._motionEval = super.createMotionEval(motion);
+        this._timelineStatsDirty = true;
+    }
+
+    public setTime(time: number) {
+        this._time = time;
```
Comment (SantyWang) [inline editor/src/marionette/preview.ts:128]: should set timeline stats dirty?

File: editor/src/marionette/preview.ts (hunk 1 above)
Comment (shrinktofit) [reply editor/src/marionette/preview.ts]: It shall not take effect now in 7a7bb82049ffe5b5eab6695a2f1a7ad6bda1879d

File: editor/src/marionette/preview.ts (hunk 2 above)
Comment (shrinktofit) [reply editor/src/marionette/preview.ts]: Solved in 7a7bb82049ffe5b5eab6695a2f1a7ad6bda1879d

File: editor/src/marionette/preview.ts (hunk 4 above)
Comment (shrinktofit) [reply editor/src/marionette/preview.ts:128]: Time does not belong to timeline stats as suggested in https://github.com/shrinktofit/engine/blob/AnimationGraphEditorAPI/editor/src/marionette/preview.ts#L161

File: editor/src/marionette/preview.ts (hunk 3 above)
Comment (shrinktofit) [reply editor/src/marionette/preview.ts:94]: It's a generator whose usage for example is:

```ts
const [[id, weight]] = queryWeights(); // Get the first motion's id and weight
for (const [id, weight] of queryWeights()) { // Iterate motion weights
  if (id === 'a') {
     break;
  }
}
```

Generators save us from array construction/concatenation e.g.

**Review by SantyWang — APPROVED** [review]

**Pull request merged**